### PR TITLE
feat(server,web): Delete and restore user from the admin portal

### LINKED ISSUE #695: [Feature]: Admin - Delete Users
### Feature Details:
- [ ] In user management on the web interface, add option to delete user

### Platform
Server & Web

### PR DESCRIPTION
**Pull request by zkhan93**
implementation as discussed in #695

UI and API changes for soft deleting a user
 - added a column in userEntity deletedAt
 - added endpoints to delete and restore user

with this PR deleting a user will disable the users login and set a timestamp on user which can be used by the job to do the actual deleting

added user delete job to run at 11 PM (before all the other jobs to reduce the work other jobs might have to do, when a user is deleted)

### CODE REVIEW

**Comment by zkhan93** [comment]
https://user-images.githubusercontent.com/7071946/200199663-6e01b100-82a1-423f-b668-b85a8afcf971.mov

File: server/apps/immich/src/api-v1/user/user.service.ts [hunk 1]
```diff
@@ -105,6 +105,48 @@ export class UserService {
     }
   }
 
+  async deleteUser(authUser: AuthUserDto, userId: string): Promise<UserResponseDto> {
+    const requestor = await this.userRepository.get(authUser.id);
+    if (!requestor) {
+      throw new NotFoundException('Requestor not found');
```
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.service.ts:111]: I think this should be a unauthorized exception

File: server/apps/immich/src/api-v1/user/user.service.ts [hunk 2]
```diff
@@ -105,6 +105,48 @@ export class UserService {
...
+    }
+    if (!requestor.isAdmin) {
+      throw new BadRequestException('Unauthorized');
```
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.service.ts:114]: And this should be forbidden

File: server/apps/immich/src/api-v1/user/user.service.ts [hunk 3]
```diff
@@ -105,6 +105,48 @@ export class UserService {
...
+    }
+    const user = await this.userRepository.get(userId);
+    if (!user) {
+      throw new NotFoundException('User not found');
```
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.service.ts:118]: And this should be bad request

**Review by jrasm91 — COMMENTED** [review]
This seems like a good approach to user removal. The cron job would be via the nest module, right?

**Comment by zkhan93** [comment]
> This seems like a good approach to user removal. The cron job would be via the nest module, right?

yes I'll add that as different commit in the same PR

**Comment by zkhan93** [comment]
@alextran1502  @jrasm91 I was wondering should we enable delete cascade on `exifInfo` and `smartInfo` columns of `AssetEntity` ?

**Comment by alextran1502** [comment]
I think it is already there. If you delete an asset, all those information will also get deleted

**Comment by zkhan93** [comment]
> I think it is already there. If you delete an asset, all those information will also get deleted

hmm okay, is it automatically assumed for OneToOne Fields ? typeorm docs does not specify that explicitly, though it does makes sense if its defaulted to cascade.

**Comment by jrasm91** [comment]
This looks good. I think it makes sense to have the 7 days be configurable. Why don't we use this library for millisecond conversion?

https://github.com/vercel/ms

We could take input in this format and it could default to "7d"?

File: server/apps/immich/src/api-v1/user/response-dto/user-response.dto.ts [hunk 4]
```diff
@@ -21,5 +22,6 @@ export function mapUser(entity: UserEntity): UserResponseDto {
     profileImagePath: entity.profileImagePath,
     shouldChangePassword: entity.shouldChangePassword,
     isAdmin: entity.isAdmin,
+    deletedAt: entity.deletedAt == undefined ? null : entity.deletedAt,
```
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/response-dto/user-response.dto.ts:25]: Can this been `deletedAt: entity.deletedAt || null`?

File: server/apps/immich/src/api-v1/user/user-repository.ts [hunk 5]
```diff
@@ -88,6 +90,17 @@ export class UserRepository implements IUserRepository {
     return this.userRepository.save(user);
   }
 
+  async delete(user: UserEntity): Promise<UserEntity> {
+    if (user.isAdmin) {
+      throw new BadRequestException('Cannot delete admin user! stay sane!');
```
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user-repository.ts:95]: Lol.

File: server/apps/immich/src/api-v1/user/user.service.ts [hunk 6]
```diff
@@ -105,6 +107,48 @@ export class UserService {
...
+    const requestor = await this.userRepository.get(authUser.id);
+    if (!requestor) {
+      throw new NotFoundException('Requestor not found');
+    }
+    if (!requestor.isAdmin) {
+      throw new BadRequestException('Unauthorized');
+    }
+    const user = await this.userRepository.get(userId, true);
+    if (!user) {
+      throw new NotFoundException('User not found');
```
Comment (jrasm91) [inline server/apps/immich/src/api-v1/user/user.service.ts:141]: Can you change these to match the ones above?

File: server/apps/immich/src/modules/schedule-tasks/schedule-tasks.service.ts [hunk 7]
```diff
@@ -128,4 +138,14 @@ export class ScheduleTasksService {
       }
     }
   }
+
+  @Cron(CronExpression.EVERY_DAY_AT_11PM)
+  async deleteUserAndRelatedAssets() {
```
Comment (jrasm91) [inline server/apps/immich/src/modules/schedule-tasks/schedule-tasks.service.ts:143]: Looks good!

File: server/apps/microservices/src/processors/user-deletion.processor.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  @Process(userDeletionProcessorName)
+  async processUserDeletion(job: Job<IUserDeletionJob>) {
+    const { user } = job.data;
+    const userAssets = await this.assetRepository.find({ where: { userId: user.id } });
+    // just for extra protection here
+    if (userUtils.isReadyForDeletion(user)) {
+      for (const asset of userAssets) {
+        assetUtils.deleteFiles(asset);
+        this.assetRepository.remove(asset);
+      }
```
Comment (jrasm91) [inline server/apps/microservices/src/processors/user-deletion.processor.ts:30]: This is technically fine, but it does bring up some good questions if the user has added a lot of assets. I'm not sure what impact it has deleting one at a time vs bulk deleting say 10k photos, for example.

Would it make more sense to do something like `assetRepository.delete({ where: { userId }})` and potentially `fs.rmSync(dir, { recursive: true, force: true });`?

Maybe I'll ask this on discord.

File: server/libs/common/src/utils/asset-utils.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        if (err) {
+          console.log('error deleting ', asset.encodedVideoPath);
+        }
+      });
+    }
+  };
+  return { deleteFiles };
+}
+
+export const assetUtils = createAssetUtils();
```
Comment (jrasm91) [inline server/libs/common/src/utils/asset-utils.ts:42]: I don't think you need this wrapper. Can you just have `export const assetUtils = `{ deleteFiles };`?

**Review by jrasm91 — APPROVED** [review]
Added a few (semantic) comments, but looks good to me. I'd be fine with merging this with 7 day delete window, and having a separate PR to make it configurable.

**Comment by bo0tzz** [comment]
I see there's a bunch of unrelated formatting changes. Can you remove those to avoid cluttering the git history?

File: server/apps/microservices/src/processors/user-deletion.processor.ts (hunk 8 above)
Comment (zkhan93) [reply server/apps/microservices/src/processors/user-deletion.processor.ts:30]: Yes, I agree should have done that, will update the PR with this approach

File: server/libs/common/src/utils/asset-utils.ts (hunk 9 above)
Comment (zkhan93) [reply server/libs/common/src/utils/asset-utils.ts:42]: I just followed the pattern from timeUtils 😁, will change it to suggested way.

**Comment by zkhan93** [comment]
Yeah those were caused by 'npm run format' on server. Will remove the files with no actual change

**Comment by zkhan93** [comment]
> This looks good. I think it makes sense to have the 7 days be configurable. Why don't we use this library for millisecond conversion?
> 
> https://github.com/vercel/ms
> 
> We could take input in this format and it could default to "7d"? 
> 

Let's do that in a follow up PR

**Comment by bo0tzz** [comment]
When a user is marked as (soft) deleted, they should probably be prevented from logging in as well.

File: server/apps/microservices/src/processors/user-deletion.processor.ts (hunk 8 above)
Comment (jrasm91) [reply server/apps/microservices/src/processors/user-deletion.processor.ts:30]: Hey, we had a conversation in discord about this, you're good with this approach for now. I think we're planning on refactoring some of the asset delete logic in the near future to use queues, especially as we introduce trash can feature for assets.

**Comment by jrasm91** [comment]
> When a user is marked as (soft) deleted, they should probably be prevented from logging in as well.

I think this "just works", since the user is soft deleted, so it won't be returned from the user repository query, unless you pass "withDeletes".

**Comment by zkhan93** [comment]
> > When a user is marked as (soft) deleted, they should probably be prevented from logging in as well.
> 
> I think this "just works", since the user is soft deleted, so it won't be returned from the user repository query, unless you pass "withDeletes".

yep, tested this as well, this magic is because I used `DeleteDateColumn` from Typeorm

File: server/apps/microservices/src/processors/user-deletion.processor.ts (hunk 8 above)
Comment (zkhan93) [reply server/apps/microservices/src/processors/user-deletion.processor.ts:30]: made the changes anyway

**Comment by jrasm91** [comment]
Future asset delete logic to be tracked in #937.

**Comment by alextran1502** [comment]
I got this error when trying to test the PR

image

**Comment by alextran1502** [comment]
> > > When a user is marked as (soft) deleted, they should probably be prevented from logging in as well.
> > 
> > 
> > I think this "just works", since the user is soft deleted, so it won't be returned from the user repository query, unless you pass "withDeletes".
> 
> yep, tested this as well, this magic is because I used `DeleteDateColumn` from Typeorm

This is amazing but confused me at first. Damn ORM 😝

**Review by jrasm91 — APPROVED** [review]

**Pull request merged**